File: 67256.cpp
#include <string>
#include <vector>

using namespace std;

struct position{
    int row = 0;
    int col = 0;
    
    position(int number) {
        switch(number) {
            case 0:
                row = 3;
                col = 1;
                break;
            case -1:
                row = 3;
                col = 0;
                break;
            case -2:
                row = 3;
                col = 2;
                break;
            default:
                row = int((number - 1) / 3);
                col = (number - 1) % 3;
                break;
        }
    }
};


int distanceBetweenTwoPosition(position& a, position& b) {
    return abs(a.row - b.row) + abs(a.col - b.col);
}

string solution(vector<int> numbers, string hand) {
    position left = position(-1);
    position right = position(-2);
    
    string answer = "";
    
    for(auto number : numbers) {
        position target = position(number);
        if(number == 1 || number == 4 || number == 7) {
            left = target;
            answer += 'L';
        } else if(number == 3 || number == 6 || number == 9) {
            right = target;
            answer += 'R';
        } else {
            int distanceFromLeft = distanceBetweenTwoPosition(left, target);
            int distanceFromRight = distanceBetweenTwoPosition(right, target);
            
            if(distanceFromLeft < distanceFromRight || ((distanceFromLeft == distanceFromRight) && hand == "left")) {
                left = target;
                answer += 'L';
            } else if(distanceFromLeft > distanceFromRight || ((distanceFromLeft == distanceFromRight) && hand == "right")) {
                right = target;
                answer += 'R';
            }
        }
    }
    
    return answer;
}

File: 67256.swift
import Foundation

struct Coordinate {
    let row: Int
    let col: Int
    
    init(_ row: Int, _ col: Int) {
        self.row = row
        self.col = col
    }
    
    func distanceTo(_ another: Coordinate) -> Int {
        return abs(self.row - another.row) + abs(self.col - another.col)
    }
}

func solution(_ numbers:[Int], _ hand:String) -> String {
    let keyPad: [Int: Coordinate] = [
        1: Coordinate(0, 0), 2: Coordinate(0, 1), 3: Coordinate(0, 2),
        4: Coordinate(1, 0), 5: Coordinate(1, 1), 6: Coordinate(1, 2),
        7: Coordinate(2, 0), 8: Coordinate(2, 1), 9: Coordinate(2, 2),
        0: Coordinate(3, 1)
    ]
    
    var leftThumb = Coordinate(3, 0)
    var rightThumb = Coordinate(3, 2)
    
    return numbers.reduce("") {
        guard let coordinate = keyPad[$1] else { return $0 + "" }
        let leftThumbDistance = coordinate.distanceTo(leftThumb)
        let rightThumbDistance = coordinate.distanceTo(rightThumb)
        if coordinate.col == 0 || (coordinate.col == 1 && (leftThumbDistance < rightThumbDistance || (leftThumbDistance == rightThumbDistance && hand == "left"))) {
            leftThumb = coordinate
            return $0 + "L"
        } else {
            rightThumb = coordinate
            return $0 + "R"
        }
    }
}
